File: yokozawa0701/gilded_rose.rb
# frozen_string_literal: true

class GildedRose
  def initialize(items)
    @items = items
  end

  def update_quality
    @items.map do |item|
      item_factory(item).update
    end
  end

  private

  def item_factory(item)
    case item.name
    when 'Sulfuras, Hand of Ragnaros'
      Sulfuras.new(name: item.name, sell_in: item.sell_in, quality: item.quality)
    when 'Aged Brie'
      AgedBrie.new(name: item.name, sell_in: item.sell_in, quality: item.quality)
    when 'Backstage passes to a TAFKAL80ETC concert'
      Backstage.new(name: item.name, sell_in: item.sell_in, quality: item.quality)
    else
      ConcreteItem.new(name: item.name, sell_in: item.sell_in, quality: item.quality)
    end
  end
end

class Item
  attr_accessor :name, :sell_in, :quality

  def initialize(name, sell_in, quality)
    @name = name
    @sell_in = sell_in
    @quality = quality
  end

  def to_s
    "#{@name}, #{@sell_in}, #{@quality}"
  end
end

class ConcreteItem < Item
  def initialize(name:, sell_in:, quality:)
    super(name, sell_in, quality)
  end

  def update
    self.quality += calc_quality_by_normal
    self.sell_in -= 1
    self.quality += calc_quality_sell_in_negative
    self
  end

  private

  def quality_less_than_50?
    quality < 50
  end

  def calc_quality_by_normal
    quality.positive? ? -1 : 0
  end

  def calc_quality_sell_in_negative
    return 0 unless sell_in.negative?

    quality.positive? ? -1 : 0
  end
end

class Sulfuras < ConcreteItem
  def update
    self
  end
end

class AgedBrie < ConcreteItem
  def calc_quality_by_normal
    quality_less_than_50? ? 1 : 0
  end

  def calc_quality_sell_in_negative
    return 0 unless sell_in.negative?

    quality_less_than_50? ? 1 : 0
  end
end

class Backstage < ConcreteItem
  def calc_quality_by_normal
    result = 0
    result += 1 if quality_less_than_50?
    result += 1 if sell_in < 11
    result += 1 if sell_in < 6
    result
  end

  def calc_quality_sell_in_negative
    return 0 unless sell_in.negative?

    -quality
  end
end

File: etoal83/src/gildedrose.rs
use std::cmp::min;
use std::fmt::{self, Display};
pub struct Item {
    pub name: String,
    pub sell_in: i32,
    pub quality: i32,
}

impl Item {
    pub fn new(name: impl Into<String>, sell_in: i32, quality: i32) -> Item {
        Item {
            name: name.into(),
            sell_in,
            quality,
        }
    }

    pub fn next_day(&mut self) {
        match &*self.name {
            "Sulfuras, Hand of Ragnaros" => (),
            "Aged Brie" => {
                self.quality = min(self.quality + 1, 50);
                self.sell_in -= 1;
            }
            "Backstage passes to a TAFKAL80ETC concert" => {
                let increment = match self.sell_in {
                    1..=5 => 3,
                    6..=10 => 2,
                    _ => 1,
                };
                self.quality = if self.sell_in <= 0 { 0 } else { self.quality + increment };
                self.sell_in -= 1;
            }
            _ => {
                self.quality -= 1;
                self.sell_in -= 1;
            }
        };
    }
}

impl Display for Item {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}, {}, {}", self.name, self.sell_in, self.quality)
    }
}

pub struct GildedRose {
    pub items: Vec<Item>,
}

impl GildedRose {
    pub fn new(items: Vec<Item>) -> GildedRose {
        GildedRose { items }
    }

    pub fn update_quality(&mut self) {
        for item in &mut self.items {
            item.next_day();
        }
    }
}

#[cfg(test)]
mod tests {
    use super::{GildedRose, Item};
    use rstest::*;
    use speculate::speculate;

    speculate! {
        describe "Decrease sell_in and quality of an Item in next day" {
            #[fixture]
            fn rose_foo() -> GildedRose {
                let items = vec![Item::new("foo", 1, 1)];
                GildedRose::new(items)
            }

            #[rstest]
            fn decrease_positive_sell_in_by_1(mut rose_foo: GildedRose) {
                rose_foo.update_quality();
                assert_eq!(rose_foo.items[0].sell_in, 0);
            }

            #[rstest]
            fn decrease_quality_by_1(mut rose_foo: GildedRose) {
                rose_foo.update_quality();
                assert_eq!(rose_foo.items[0].quality, 0);
            }
        }

        describe "Sulfuras is never sold nor degraded" {
            #[rstest]
            fn keep_sulfuras_constant() {
                let items = vec![
                    Item::new("Sulfuras, Hand of Ragnaros", 0, 80),
                    Item::new("Sulfuras, Hand of Ragnaros", -1, 80)
                ];
                let mut rose = GildedRose::new(items);
                rose.update_quality();
                assert_eq!((rose.items[0].sell_in, rose.items[0].quality), (0, 80));
                assert_eq!((rose.items[1].sell_in, rose.items[1].quality), (-1, 80));
            }
        }
    }
}
